ensembl_genes/Red.py

```diff
@@ -0,0 +1,255 @@
+# Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
+# Copyright [2016-2020] EMBL-European Bioinformatics Institute
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Module to run Red to find repeats and store them in the given Ensembl core database """
+
+import os
+import errno
+import subprocess
+import tempfile
+import shutil
+import sqlalchemy as db
+from pathlib import Path,PurePath
+
+# sqlalchemy requires MySQLdb but MySQLdb doesn't support Python 3.x
+# pymysql can be imported and used instead
+import pymysql
+pymysql.install_as_MySQLdb()
+
+import eHive
+
+class Red(eHive.BaseRunnable):
+    """ Runnable that runs Red to find repeats and store them in the target database. """
+
+    def param_defaults(self):
+        """ It sets the parameters default values. """
+        return {
+            'logic_name' : 'red',
+            'target_db_url' : '', # 'driver://user:pass@host:port/dbname'
+            'red_path' : '',
+        }
+
+    def fetch_input(self):
+        """ It fetches the input parameters and it checks that they are correct. """
+
+        # get new temporary directory name in the default path with prefix gnm_ (eg '/scratch/gnm_nar4nry8')
+        temp_gnm_dir = tempfile.TemporaryDirectory(prefix='gnm_').name
+
+        genome_file = self.param_required('genome_file')
+        gnm = self.param('gnm',temp_gnm_dir)
+        msk = self.param_required('msk')
+        rpt = self.param_required('rpt')
+        red_path = self.param_required('red_path')
+        target_db_url = self.param_required('target_db_url')
+        self.param('target_db_url',f'{target_db_url}?local_infile=1')
+
+        # make the temporary directory 'gnm' and copy the genome file into it
+        # in this way we make sure that the only .fa file to be processed is the one we want
+        try:
+            Path(gnm).mkdir()
+        except PermissionError:
+            print(f'Could not create {gnm} directory.')
+            raise
+
+        # copy the genome file into the temporary directory
+        # add suffix '.fa' to make sure it ends with '.fa' as required by Red
+        # In this way we make sure that the only .fa file to be processed is the one we want
+        new_genome_file = Path(f'{gnm}/{PurePath(genome_file).name}.fa')
+        try:
+            shutil.copyfile(genome_file,new_genome_file)
```

Comment:
I was thinking, instead of copying, could we do a symlink? It would not help much for a single assembly but if we have tens of pipeline running, it would help the filesystem

Reply:
Good idea. Just tested and pushed. Thank you.